### PR TITLE
fix csdb index error

### PR DESCRIPTION
**Pull request by noot**
Closes: #491 

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/types/statedb.go [hunk 1]
```diff
@@ -702,38 +702,39 @@ func (csdb *CommitStateDB) Copy() *CommitStateDB {
...
+			state.addressToObjectIndex[dirty.address] = len(state.stateObjects) - 1
 		}
 	}
 
 	// Above, we don't copy the actual journal. This means that if the copy is
 	// copied, the loop above will be a no-op, since the copy's journal is empty.
 	// Thus, here we iterate over stateObjects, to enable copies of copies.
 	for addr := range csdb.stateObjectsDirty {
-		if idx, exist := state.addressToObjectIndex[addr]; !exist {
+		if idx, exist := state.addressToObjectIndex[addr]; exist {
```
Comment (fedekunze) [inline x/evm/types/statedb.go]: I think this should be `!exist`?

File: x/evm/types/statedb.go [hunk 2]
```diff
@@ -702,38 +702,39 @@ func (csdb *CommitStateDB) Copy() *CommitStateDB {
...
 		if idx, exist := csdb.addressToObjectIndex[dirty.address]; exist {
-			state.stateObjects[i] = stateEntry{
+			state.stateObjects = append(state.stateObjects, stateEntry{
 				address:     dirty.address,
 				stateObject: csdb.stateObjects[idx].stateObject.deepCopy(state),
-			}
-			state.stateObjectsDirty[dirty.address] = struct{}{}
+			})
+			delete(state.stateObjectsDirty, dirty.address)
+			state.addressToObjectIndex[dirty.address] = len(state.stateObjects) - 1
```
Comment (fedekunze) [inline x/evm/types/statedb.go:727]: this is the fix, right?

File: x/evm/types/statedb.go (hunk 1 above)
Comment (noot) [reply x/evm/types/statedb.go]: I think so but we are using `idx` in the line right after, which is bad if it doesn't exist

File: x/evm/types/statedb.go (hunk 2 above)
Comment (noot) [reply x/evm/types/statedb.go:727]: unfortunately not, the fix seemed to be `idx < len(csdb.stateObjects)` inside `getStateObject` and `setStateObject` which I don't really like...

File: x/evm/types/statedb.go (hunk 1 above)
Comment (fedekunze) [reply x/evm/types/statedb.go]: https://github.com/ChainSafe/ethermint/pull/352/files#diff-3a1f916b2ed8cadd721b0cf84b47b62fL693-L701

Comment (noot) [reply x/evm/types/statedb.go]: i see, will update!

**Review by fedekunze — APPROVED** [review]
utACK. Will look into the other test failures tomorrow

**Pull request merged**